### PR TITLE
Storage monitor added to polkadot node

### PR DESCRIPTION
**Pull request by michalkucharczyk**
Related to: https://github.com/paritytech/substrate/issues/12399

### CODE REVIEW

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2307847

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2309006

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2310901

**Comment by michalkucharczyk** [comment]
bot rebase

**Comment by michalkucharczyk** [comment]
Initial check doesn't work well, due to some endless, busyloop in network components.
```
2023-01-25 12:51:23.485  INFO main sc_cli::runner: Parity Polkadot    
2023-01-25 12:51:23.485  INFO main sc_cli::runner: ✌️  version 0.9.37-6139c1af8e9    
2023-01-25 12:51:23.485  INFO main sc_cli::runner: ❤️  by Parity Technologies <admin@parity.io>, 2017-2023    
2023-01-25 12:51:23.485  INFO main sc_cli::runner: 📋 Chain specification: Polkadot    
2023-01-25 12:51:23.485  INFO main sc_cli::runner: 🏷  Node name: mature-cap-4570    
2023-01-25 12:51:23.485  INFO main sc_cli::runner: 👤 Role: FULL    
2023-01-25 12:51:23.485  INFO main sc_cli::runner: 💾 Database: RocksDb at /home/miszka/.local/share/polkadot/chains/polkadot/db/full    
2023-01-25 12:51:23.485  INFO main sc_cli::runner: ⛓  Native runtime: polkadot-9370 (parity-polkadot-0.tx20.au0)    
...
2023-01-25 12:51:28.214 DEBUG                 main storage-monitor: Initializing StorageMonitorService for db path: "/home/miszka/.local/share/polkadot/chains/polkadot/db/full"    
2023-01-25 12:51:28.214 TRACE tokio-runtime-worker sub-libp2p: PSM => Connect(12D3KooWSz8r2WyCdsfWHgPyvD8GKQdJ1UAiRmrcrs8sQB3fe2KU, SetId(0)): Starting to connect    
2023-01-25 12:51:28.214 TRACE                 main storage-monitor: free: 500429 , threshold: 1000000.    
2023-01-25 12:51:28.214 TRACE tokio-runtime-worker sub-libp2p: Libp2p <= Dial 12D3KooWSz8r2WyCdsfWHgPyvD8GKQdJ1UAiRmrcrs8sQB3fe2KU    
2023-01-25 12:51:28.214 ERROR                 main storage-monitor: Available space 500429MB for path `/home/miszka/.local/share/polkadot/chains/polkadot/db/full` dropped below threshold: 1000000MB , terminating...    
...
2023-01-25 12:51:28.214 DEBUG tokio-runtime-worker trust_dns_resolver::name_server::name_server_pool: sending request: [Query { name: Name("telemetry.polkadot.io"), query_type: A, query_class: IN, mdns_unicast_response: false }]
2023-01-25 12:51:28.214 DEBUG tokio-runtime-worker trust_dns_resolver::name_server::name_server: reconnecting: NameServerConfig { socket_addr: 127.0.0.1:53, protocol: Udp, tls_dns_name: None, trust_nx_responses: false, bind_addr: None }
2023-01-25 12:51:28.214 DEBUG tokio-runtime-worker trust_dns_resolver::name_server::name_server: name_server connection failure: proto error: resource too busy
2023-01-25 12:51:28.214 TRACE tokio-runtime-worker telemetry: Dialing    
2023-01-25 12:51:28.214 TRACE tokio-runtime-worker telemetry: Dialing    
...
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: polling Connection    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::connection: reading incoming messages    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: poll_read_messages called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: polling socket    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::codecs: NetlinkCodec: decoding next message    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::codecs: buffer is empty    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_sys::tokio: poll_recv_from called    
2023-01-25 12:51:28.215 ERROR tokio-runtime-worker netlink_proto::framed: failed to read from netlink socket: Custom { kind: Other, error: "A Tokio 1.x context was found, but it is being shutdown." }    
2023-01-25 12:51:28.215  WARN tokio-runtime-worker netlink_proto::connection: netlink socket stream shut down    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::connection: forwarding unsolicited messages to the connection handle    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: forward_unsolicited_messages called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: forward_unsolicited_messages done    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::connection: forwaring responses to previous requests to the connection handle    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: forward_responses called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: forward_responses done    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::connection: handling requests    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: poll_requests called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: no new requests to handle poll_requests done    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::connection: sending messages    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: poll_send_messages called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: found outgoing message to send checking if socket is ready    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: sending outgoing message    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::framed: sending frame    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::codecs: >>> NetlinkMessage { header: NetlinkHeader { length: 24, message_type: 22, flags: 769, sequence_number: 1, port_number: 0 }, payload: InnerMessage(GetAddress(AddressMessage { header: AddressHeader { family: 0, prefix_len: 0, flags: 0, scope: 0, index: 0 }, nlas: [] })) }    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::framed: frame encoded; length=24    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: poll_send_messages done    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: poll_flush called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::framed: flushing frame; length=24    
2023-01-25 12:51:28.215  WARN tokio-runtime-worker netlink_proto::connection: error flushing netlink socket: Custom { kind: Other, error: "A Tokio 1.x context was found, but it is being shutdown." }    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: done polling Connection    
2023-01-25 12:51:28.215 ERROR tokio-runtime-worker libp2p_mdns::behaviour: if watch returned an error: rtnetlink socket closed    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: polling Connection    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::connection: reading incoming messages    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: poll_read_messages called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: polling socket    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::codecs: NetlinkCodec: decoding next message    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::codecs: buffer is empty    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_sys::tokio: poll_recv_from called    
2023-01-25 12:51:28.215 ERROR tokio-runtime-worker netlink_proto::framed: failed to read from netlink socket: Custom { kind: Other, error: "A Tokio 1.x context was found, but it is being shutdown." }    
2023-01-25 12:51:28.215  WARN tokio-runtime-worker netlink_proto::connection: netlink socket stream shut down    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::connection: forwarding unsolicited messages to the connection handle    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: forward_unsolicited_messages called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: forward_unsolicited_messages done    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::connection: forwaring responses to previous requests to the connection handle    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: forward_responses called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: forward_responses done    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::connection: handling requests    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: poll_requests called    
2023-01-25 12:51:28.215 DEBUG tokio-runtime-worker netlink_proto::connection: sending messages    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: poll_send_messages called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: poll_send_messages done    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: poll_flush called    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::framed: flushing frame; length=24    
2023-01-25 12:51:28.215  WARN tokio-runtime-worker netlink_proto::connection: error flushing netlink socket: Custom { kind: Other, error: "A Tokio 1.x context was found, but it is being shutdown." }    
2023-01-25 12:51:28.215 TRACE tokio-runtime-worker netlink_proto::connection: done polling Connection    
2023-01-25 12:51:28.215 ERROR tokio-runtime-worker libp2p_mdns::behaviour: if watch returned an error: rtnetlink socket closed    
...
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: polling Connection    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::connection: reading incoming messages    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: poll_read_messages called    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: polling socket    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::codecs: NetlinkCodec: decoding next message    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::codecs: buffer is empty    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_sys::tokio: poll_recv_from called    
2023-01-25 12:51:28.999 ERROR tokio-runtime-worker netlink_proto::framed: failed to read from netlink socket: Custom { kind: Other, error: "A Tokio 1.x context was found, but it is being shutdown." }    
2023-01-25 12:51:28.999  WARN tokio-runtime-worker netlink_proto::connection: netlink socket stream shut down    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::connection: forwarding unsolicited messages to the connection handle    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: forward_unsolicited_messages called    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: forward_unsolicited_messages done    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::connection: forwaring responses to previous requests to the connection handle    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: forward_responses called    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: forward_responses done    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::connection: handling requests    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: poll_requests called    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::connection: sending messages    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: poll_send_messages called    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: poll_send_messages done    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: poll_flush called    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::framed: flushing frame; length=24    
2023-01-25 12:51:28.999  WARN tokio-runtime-worker netlink_proto::connection: error flushing netlink socket: Custom { kind: Other, error: "A Tokio 1.x context was found, but it is being shutdown." }    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: done polling Connection    
2023-01-25 12:51:28.999 ERROR tokio-runtime-worker libp2p_mdns::behaviour: if watch returned an error: rtnetlink socket closed    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: polling Connection    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::connection: reading incoming messages    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: poll_read_messages called    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: polling socket    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::codecs: NetlinkCodec: decoding next message    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::codecs: buffer is empty    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_sys::tokio: poll_recv_from called    
2023-01-25 12:51:28.999 ERROR tokio-runtime-worker netlink_proto::framed: failed to read from netlink socket: Custom { kind: Other, error: "A Tokio 1.x context was found, but it is being shutdown." }    
2023-01-25 12:51:28.999  WARN tokio-runtime-worker netlink_proto::connection: netlink socket stream shut down    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::connection: forwarding unsolicited messages to the connection handle    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: forward_unsolicited_messages called    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: forward_unsolicited_messages done    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::connection: forwaring responses to previous requests to the connection handle    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: forward_responses called    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: forward_responses done    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::connection: handling requests    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: poll_requests called    
2023-01-25 12:51:28.999 DEBUG tokio-runtime-worker netlink_proto::connection: sending messages    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: poll_send_messages called    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: poll_send_messages done    
2023-01-25 12:51:28.999 TRACE tokio-runtime-worker netlink_proto::connection: poll_flush called    
2023-01-25 12:51:29.000 TRACE tokio-runtime-worker netlink_proto::framed: flushing frame; length=24
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2320412

**Comment by michalkucharczyk** [comment]
bot rebase

File: cli/src/command.rs [hunk 1]
```diff
@@ -349,7 +350,16 @@ where
 			hwbench,
 		)
 		.map(|full| full.task_manager)
-		.map_err(Into::into)
+		.map_err(Into::<Error>::into)?;
```
Comment (bkchr) [inline cli/src/command.rs:353]:
```suggestion
```

File: cli/src/command.rs [hunk 2]
```diff
@@ -349,7 +350,16 @@ where
 			hwbench,
 		)
 		.map(|full| full.task_manager)
```
Comment (bkchr) [inline cli/src/command.rs:352]:
```suggestion
		.map(|full| full.task_manager)?;
```

`?` is calling `From::from` automatically. This being the reason you need to write `Into::<Error>`

**Review by bkchr — APPROVED** [review]

File: cli/src/command.rs [hunk 3]
```diff
@@ -349,7 +350,16 @@ where
...
-		.map_err(Into::into)
+		.map_err(Into::<Error>::into)?;
+
+		sc_storage_monitor::StorageMonitorService::try_spawn(
+			cli.storage_monitor,
+			database_source,
+			&task_manager.spawn_essential_handle(),
+		)
+		.map_err(Into::<Error>::into)?;
```
Comment (bkchr) [inline cli/src/command.rs:360]:
```suggestion
		)?;
```

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2322562

File: cli/src/cli.rs [hunk 4]
```diff
@@ -151,6 +151,11 @@ pub struct RunCmd {
 pub struct Cli {
 	#[command(subcommand)]
 	pub subcommand: Option<Subcommand>,
+
 	#[clap(flatten)]
 	pub run: RunCmd,
+
+	#[allow(missing_docs)]
```
Comment (davxy) [inline cli/src/cli.rs:158]:
```suggestion
```
The `allow(missing_docs)` directive on the struct should cover its members as well

**Review by davxy — APPROVED** [review]

File: cli/src/cli.rs (hunk 4 above)
Comment (michalkucharczyk) [reply cli/src/cli.rs]: https://github.com/paritytech/polkadot/pull/6619/commits/4cea6fed2ee475a65db81d0726a5fdd6ba5ed3e7

**Comment by paritytech-cicd-pr** [comment]
The CI pipeline was cancelled due to failure one of the required jobs.
**Job name**: test-linux-stable
**Logs**: https://gitlab.parity.io/parity/mirrors/polkadot/-/jobs/2327812

**Pull request merged**